File: php/src/Webhook/WebhookEvent.php
<?php

declare(strict_types=1);

namespace Fivexer\SDK\Webhook;

use Fivexer\SDK\Exception\FivexerException;

/**
 * A verified webhook payload. event is the event type (e.g. 'task.matched')
 * when the payload carried one; data is the full parsed JSON array.
 */
final class WebhookEvent
{
    /**
     * @param array<string, mixed>|null $data
     */
    public function __construct(
        public readonly ?string $event,
        public readonly ?array $data,
    ) {
    }

    /**
     * @throws FivexerException when the payload is not valid JSON
     */
    public static function fromPayload(string $payload): self
    {
        $decoded = \json_decode($payload, true);
        if (\json_last_error() !== \JSON_ERROR_NONE) {
            throw new FivexerException('webhook payload is not valid JSON');
        }
        // Valid JSON but not an object (a JSON array or scalar): no event, no data —
        // matches the Java SDK's handling of unexpected webhook shapes.
        if (!\is_array($decoded) || (\array_is_list($decoded) && $decoded !== [])) {
            return new self(null, null);
        }
        $event = isset($decoded['event']) && \is_string($decoded['event']) ? $decoded['event'] : null;
        return new self($event, $decoded);
    }
}

File: java/src/main/java/io/fivexer/sdk/webhook/WebhookEvent.java
package io.fivexer.sdk.webhook;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import io.fivexer.sdk.FivexerException;
import java.nio.charset.StandardCharsets;
import java.util.Map;

/**
 * A verified webhook payload. {@link #getEvent()} is the event type (e.g. {@code task.matched})
 * when the payload carried one; {@link #getData()} is the full parsed JSON object.
 */
public class WebhookEvent {

    private static final Gson GSON = new Gson();

    private final String event;
    private final Map<String, Object> data;

    public WebhookEvent(String event, Map<String, Object> data) {
        this.event = event;
        this.data = data;
    }

    public String getEvent() { return event; }

    @SuppressWarnings("unchecked")
    public Map<String, Object> getData() { return data; }

    @SuppressWarnings("unchecked")
    static WebhookEvent fromPayload(byte[] payload) {
        try {
            JsonElement root = JsonParser.parseString(new String(payload, StandardCharsets.UTF_8));
            if (!root.isJsonObject()) {
                return new WebhookEvent(null, null);
            }
            JsonObject obj = root.getAsJsonObject();
            Map<String, Object> map = GSON.fromJson(obj, Map.class);
            String event = obj.has("event") && obj.get("event").isJsonPrimitive()
                    ? obj.get("event").getAsString() : null;
            return new WebhookEvent(event, map);
        } catch (Exception e) {
            throw new FivexerException("webhook payload is not valid JSON", e);
        }
    }
}
